bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+macro_rules! impl_div_mul {
+    ($ty:ident, $int:ident) => {
+        impl core::ops::Div<$int> for $ty {
+            type Output = $ty;
+            #[inline]
+            fn div(self, rhs: $int) -> Self {
+                let rhs = U256::from(u64::from(rhs));
+                $ty(self.0 / rhs)
+            }
+        }
+
+        impl Mul<$int> for $ty {
+            type Output = $ty;
+            #[inline]
+            fn mul(self, rhs: $int) -> Self { $ty(self.0.mul(rhs)) }
+        }
+
+        impl Mul<$ty> for $int {
+            type Output = $ty;
+            #[inline]
+            fn mul(self, rhs: $ty) -> $ty { $ty(self.mul(rhs.0)) }
+        }
+    };
+}
+
+// Implement traits and methods shared by `Target` and `Work`.
+macro_rules! do_impl {
+    ($ty:ident) => {
+        impl $ty {
+            /// Creates `Self` from a big-endian byte array.
+            #[inline]
+            pub fn from_be_bytes(bytes: [u8; 32]) -> $ty { $ty(U256::from_be_bytes(bytes)) }
+
+            /// Creates `Self` from a little-endian byte array.
+            #[inline]
+            pub fn from_le_bytes(bytes: [u8; 32]) -> $ty { $ty(U256::from_le_bytes(bytes)) }
+
+            /// Converts `self` to a big-endian byte array.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+
+            /// Converts `self` to a little-endian byte array.
+            #[inline]
+            pub fn to_le_bytes(self) -> [u8; 32] { self.0.to_le_bytes() }
+        }
+
+        impl_from_unsigned_type!($ty, u8);
+        impl_from_unsigned_type!($ty, u16);
+        impl_from_unsigned_type!($ty, u32);
+        impl_from_unsigned_type!($ty, u64);
+        impl_from_unsigned_type!($ty, u128);
+
+        impl_try_from_signed_type!($ty, i8);
+        impl_try_from_signed_type!($ty, i16);
+        impl_try_from_signed_type!($ty, i32);
+        impl_try_from_signed_type!($ty, i64);
+        impl_try_from_signed_type!($ty, i128);
+
+        impl fmt::Display for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
+        }
+
+        impl fmt::LowerHex for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::LowerHex::fmt(&self.0, f) }
+        }
+
+        impl fmt::UpperHex for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::UpperHex::fmt(&self.0, f) }
+        }
+
+        impl_div_mul!($ty, u8);
+        impl_div_mul!($ty, u16);
+        impl_div_mul!($ty, u32);
+        impl_div_mul!($ty, u64);
+
+        impl Shl<u32> for $ty {
+            type Output = Self;
+            #[inline]
+            fn shl(self, shift: u32) -> Self { Self(self.0 << shift) }
+        }
+
+        impl Shr<u32> for $ty {
+            type Output = Self;
+            #[inline]
+            fn shr(self, shift: u32) -> Self { Self(self.0 >> shift) }
+        }
```

Comment:
Bit operations on things that we no longer consider bags of bits are quite surprising. I'd rather remove them.

Reply:
Fair point, removed. Cheers.